core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java

```diff
@@ -122,10 +127,32 @@ public RelJson withInputTranslator(InputTranslator inputTranslator) {
     return new RelJson(jsonBuilder, inputTranslator);
   }
 
+  /** Returns a RelJson with a given RelJsonReader. */
+  @SuppressWarnings("initialization.invalid.field.write.initialized")
+  public RelJson withRelJsonReader(@UnknownInitialization RelJsonReader relJsonReader) {
+    this.relJsonReader = relJsonReader;
+    return this;
+  }
+
+  /** Returns a RelJson with a given RelJsonWriter. */
+  @SuppressWarnings("initialization.invalid.field.write.initialized")
+  public RelJson withRelJsonWriter(@UnknownInitialization RelJsonWriter relJsonWriter) {
+    this.relJsonWriter = relJsonWriter;
+    return this;
+  }
+
   private JsonBuilder jsonBuilder() {
     return requireNonNull(jsonBuilder, "jsonBuilder");
   }
 
+  private RelJsonWriter relJsonWriter() {
+    return requireNonNull(relJsonWriter, "relJsonWriter");
+  }
+
+  private RelJsonReader relJsonReader() {
```

Comment:
It's a little weird that you add RelJsonWriter and RelJsonReader to RelJson. Why RelJson has to use them?

Reply:
Because there is a `RelNode` in `RexSubQuery`, to serialize/deserialize a `RelNode`, we need to use `RelJsonWriter` and `RelJsonReader`.